escrow/atomic_storage.go

```diff
@@ -24,6 +24,29 @@ type AtomicStorage interface {
 	CompareAndSwap(key string, prevValue string, newValue string) (ok bool, err error)
 	// Delete removes value by key
 	Delete(key string) (err error)
+
+	StartTransaction(keyPrefix string) (transaction Transaction, err error)
+	CompleteTransaction(transaction Transaction, update []*KeyValueData) (ok bool, err error)
+	ExecuteTransaction(request CASRequest) (ok bool, err error)
+}
+
+type Transaction interface {
+	GetConditionValues() ([]*KeyValueData, error)
+}
+
+//Best to change this to KeyValueData , will do this in the next commit
+type UpdateFunc func(conditionValues []*KeyValueData) (update []*KeyValueData, err error)
+
+type CASRequest struct {
+	RetryTillSuccessOrError bool
+	Update                  UpdateFunc
+	ConditionKeyPrefix      string
```

Comment:
One last thing I would change in this API is replacing 
```golang
ConditionalKeyPrefix string
```
by something like 
```golang
ConditionalKeys []KeyOrPrefix

type KeyOrPrefix struct {
   Prefix bool
   Key string
}
```
or simply by
```golang
ConditionKeys []string
```
Because in particular use case this PR addresses we actually know all three keys we need and we can simplify code by using strict keys. Also this will allow removing `ChannelId` and `UsageType` fields from `PrePaidDataUnit` and making it simple:
```golang
type PrePaidDataUnit struct {
	Amount    *big.Int
}
```
It is possible because information about condition keys is now passed with condition values in `KeyValueData` structure.

Reply:
More than happy to do so  and agree :)

Reply:
got this comment addressed, on exhaustive test cases next